tools/codegen/generate.sh

```
@@ -265,7 +265,30 @@ EOF
+    else
+        find "$output_dir" -name "*.py" -type f -exec sed -i "$sed_script" {} \;
+    fi
+
+    # Strip trailing whitespace and ensure files end with single newline (pre-commit compatibility)
+    # Note: Generated files are excluded from black/isort in pre-commit config
+    log_info "Normalizing whitespace in generated files..."
+    if [[ "$(uname)" == "Darwin" ]]; then
+        # Strip trailing whitespace from lines
+        find "$output_dir" -name "*.py" -type f -exec sed -i '' 's/[[:space:]]*$//' {} \;
+        # Ensure file ends with exactly one newline
+        for f in $(find "$output_dir" -name "*.py" -type f); do
```

Shellcheck (0.11.0)

[warning] For loops over find output are fragile. Use find -exec or a while read loop.

(SC2044)